dfdaemon/config/config.go

```diff
@@ -413,3 +418,15 @@ func NewProxy(regx string, useHTTPS bool, direct bool, redirect string) (*Proxy,
 func (r *Proxy) Match(url string) bool {
 	return r.Regx != nil && r.Regx.MatchString(url)
 }
+
+// ProtocolConfig is the config of url protocol.
+type ProtocolConfig struct {
+	Name string `yaml:"name" json:"name"`
+	// Opts defines the opts for protocol, such as tls config.
+	Opts map[string]interface{} `yaml:"opts" json:"opts"`
+}
+
+type PatternConfig struct {
+	Pattern string                 `yaml:"pattern" json:"pattern"`
+	Opts    map[string]interface{} `yaml:"opts" json:"opts"`
```

Comment:
`interface{}` inside a configuration is trouble-some. It is not easy to see what it is really.

If possible, this(the type) needs to be spelled out explicitly.

Reply:
Agreed. But it seems the content of this this configuration is determined by corresponding plugins, and we're not sure what the type would be. If this was json, I would recommend using `json.RawMessage`, but there's no yaml equivalent. I a plugin, we may need to `yaml.Marshal` first then `yaml.Unmarshal` again to unmarshal this field into the specific type.

Reply:
I think maybe it's fair to say we can start with `map[string]string`.

Reply:
Or maybe something like this?

```go
type RawYAML yaml.MapSlice

func (msg RawYAML) Unmarshal(v interface{}) error {
        body, _ := yaml.Marshal(msg)
        return yaml.Unmarshal(body, v)
}
```

Then each plugin can have their own typed config:

```go
var customPluginConfig struct {
  // ...
}

config.Opts.Unmarshal(&customPluginConfig)
```

Reply:
Sounds good to me. Thanks for sharing this.

Reply:
Actually,  echo plugin could unmarshal `map[string]interface{}` into their own config. And json and yaml should be both supported in reading config, so we could not use type of json or yaml as field of opt. On this condition, `map[string]interface{}` may be a good  choice.

Reply:
@inoc603 @zhouhaibing089  PTAL